cirq/value/linear_dict.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Linear combination represented as mapping of things to coefficients."""
+
+from typing import Any, Dict, Mapping, Union
+
+Scalar = Union[complex, float]
+
+
+class LinearDict(Dict[Any, Scalar]):
+    """Represents linear combination of things.
+
+    LinearDict implements the basic linear algebraic operations of vector
+    addition and scalar multiplication for linear combinations of abstract
+    vectors. Keys represent the vectors, values represent their coefficients.
+    The only requirement on the keys is that they be hashable.
+
+    A consequence of treating keys as opaque is that all relationships between
+    the keys other than equality are ignored. In other words, we treat linear
+    combinations as if all distinct keys are independent.
+
+    For convenience, methods which take a second linear combination accept
+    a standalone key instance as well. The second argument is then treated as
+    a single-term linear combination with the sole coefficient equal to 1.0.
+    """
+    def __init__(self, terms: Mapping[Any, Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of abstract vectors to coefficients in the linear
+                combination being initialized.
+        """
+        super(LinearDict, self).__init__()
+        self.update(terms)
+
+    def clean(self, *, tolerance: float=1e-9) -> None:
+        """Modifies self by deleting terms with small coefficients."""
+        negligible = [v for v, c in self.items() if abs(c) <= tolerance]
+        for v in negligible:
+            del self[v]
+
+    def copy(self) -> 'LinearDict':
+        return LinearDict(super(LinearDict, self).copy())
+
+    def __getitem__(self, vector: Any) -> Scalar:
+        if vector in self:
```

Comment:
Rename `vector` to `key` then just `return super().get(key, 0)`

Reply:
I like `vector`. Argument name `key` just tells the reader what they already know: this is a type of dictionary so naturally `__getitem__()` takes a key as argument. OTOH, `vector` reminds them that `LinearDict` endows its keys with additional meaning: they represent vectors which are combined in the linear combination.

Agreed on the rest of your comment.

Reply:
Perhaps `basis` would be even more accurate?

Reply:
Considered it. Decided against to avoid the implication that elements of combination are linearly independent. We never check this and user should not expect it. In particular, this is legal: `LinearDict({cirq.X: 1, cirq.Z: 1, cirq.H: 1})` even though the operators are linearly dependent.

Will clarify in class docstring.

Reply:
Done: `return super().get(vector, 0)` and docstring.